File: apps/starter-ios/StarterApp/Core/Sync/SyncEngine.swift
import Foundation

@Observable
final class SyncEngine {
    var state: SyncState = .idle
    var lastSyncAt: Date?
    var unresolvedConflicts: [AnyConflict] = []

    private let repository: any SyncRepository
    private let apiClient: any SyncAPIClient
    private let network: NetworkMonitor
    private let metrics: SyncMetrics
    private let queue: SyncQueue = SyncQueue()
    private var pollingTask: Task<Void, Never>?

    private var deviceId: String {
        if let stored = UserDefaults.standard.string(forKey: "com.appstarterkit.sync.deviceId") {
            return stored
        }
        let new = UUID().uuidString
        UserDefaults.standard.set(new, forKey: "com.appstarterkit.sync.deviceId")
        return new
    }

    init(
        repository: any SyncRepository,
        apiClient: any SyncAPIClient,
        network: NetworkMonitor,
        metrics: SyncMetrics
    ) {
        self.repository = repository
        self.apiClient = apiClient
        self.network = network
        self.metrics = metrics
    }

    func sync() async {
        guard network.isConnected else {
            state = .offline
            return
        }
        guard state != .syncing else { return }

        state = .syncing
        let start = Date.now

        do {
            // Push pending local changes
            let pending = try await repository.fetchPending()
            let changes = pending.compactMap { record -> SyncChange? in
                guard let action = record.syncAction else { return nil }
                return SyncChange(
                    id: record.id,
                    collection: String(describing: type(of: record)),
                    action: action,
                    payload: [:],
                    updatedAt: ISO8601DateFormatter().string(from: record.updatedAt)
                )
            }

            var pushedCount = 0
            var conflictCount = 0
            var errorCount = 0

            if !changes.isEmpty {
                let response = try await apiClient.syncPush(changes: changes, deviceId: deviceId)
                pushedCount = response.applied.count
                errorCount = response.errors.count
                let appliedIds = Set(response.applied.map(\.id))
                queue.remove(ids: appliedIds)

                // Handle conflicts
                conflictCount = response.conflicts.count
                let newConflicts = response.conflicts.map { c in
                    AnyConflict(id: c.id, collection: c.collection, conflict: c)
                }
                unresolvedConflicts.append(contentsOf: newConflicts)
            }

            // Pull server changes since last sync
            let pullResponse = try await apiClient.syncPull(since: lastSyncAt, collections: [])
            try await repository.applyServerChanges(pullResponse.items)

            let duration = Int(Date.now.timeIntervalSince(start) * 1000)
            metrics.record(SyncRecord(
                timestamp: start,
                pushedCount: pushedCount,
                pulledCount: pullResponse.items.count,
                conflictCount: conflictCount,
                errorCount: errorCount,
                durationMs: duration,
                succeeded: true
            ))

            lastSyncAt = .now
            state = .idle
        } catch {
            let duration = Int(Date.now.timeIntervalSince(start) * 1000)
            metrics.record(SyncRecord(
                timestamp: start,
                pushedCount: 0,
                pulledCount: 0,
                conflictCount: 0,
                errorCount: 1,
                durationMs: duration,
                succeeded: false
            ))
            state = .error(error.localizedDescription)
        }
    }

    func startPolling(interval: TimeInterval = 30) {
        stopPolling()
        pollingTask = Task { [weak self] in
            while !Task.isCancelled {
                await self?.sync()
                try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
            }
        }
    }

    func stopPolling() {
        pollingTask?.cancel()
        pollingTask = nil
    }

    func resolve(_ conflict: AnyConflict, strategy: ResolutionStrategy) async {
        do {
            try await repository.resolveConflict(conflict, strategy: strategy)
            unresolvedConflicts.removeAll { $0.id == conflict.id }
        } catch {
            state = .error(error.localizedDescription)
        }
    }
}

File: apps/starter-android/app/src/main/kotlin/com/onlystack/starterapp/core/sync/SyncEngine.kt
package com.onlystack.starterapp.core.sync

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import com.onlystack.starterapp.core.network.NetworkMonitor
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.Date
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SyncEngine @Inject constructor(
    private val context: Context,
    private val network: NetworkMonitor,
    private val metrics: SyncMetrics,
) {
    var repository: SyncRepository? = null
    var apiClient: SyncApiService? = null

    private val _state = MutableStateFlow<SyncState>(SyncState.Idle)
    val state: StateFlow<SyncState> = _state

    var lastSyncAt: Date? = null
        private set

    private val _unresolvedConflicts = MutableStateFlow<List<SyncConflict>>(emptyList())
    val unresolvedConflicts: StateFlow<List<SyncConflict>> = _unresolvedConflicts

    private val queue = SyncQueue()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var pollingJob: Job? = null

    private val deviceId: String by lazy {
        val prefs = context.getSharedPreferences("sync_engine", Context.MODE_PRIVATE)
        prefs.getString("device_id", null) ?: UUID.randomUUID().toString().also { id ->
            prefs.edit().putString("device_id", id).apply()
        }
    }

    private fun isOnline(): Boolean {
        val cm = context.getSystemService(ConnectivityManager::class.java)
        val capabilities = cm.getNetworkCapabilities(cm.activeNetwork) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
            capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
    }

    suspend fun sync() {
        if (!isOnline()) {
            _state.value = SyncState.Offline
            return
        }
        if (_state.value is SyncState.Syncing) return

        val repo = repository ?: return
        val api = apiClient ?: return

        _state.value = SyncState.Syncing
        val start = System.currentTimeMillis()

        runCatching {
            val pending = repo.fetchPending()
            val changes = pending.mapNotNull { entity ->
                val action = entity.syncAction ?: return@mapNotNull null
                SyncChange(
                    id = entity.id,
                    collection = entity::class.simpleName ?: "unknown",
                    action = action,
                    payload = emptyMap(),
                    updatedAt = entity.updatedAt.toInstant().toString(),
                )
            }

            var pushedCount = 0
            var conflictCount = 0
            var errorCount = 0

            if (changes.isNotEmpty()) {
                val response = api.syncPush(SyncPushRequest(changes, deviceId))
                pushedCount = response.applied.size
                errorCount = response.errors.size
                val appliedIds = response.applied.map { it.id }.toSet()
                queue.remove(appliedIds)
                conflictCount = response.conflicts.size
                _unresolvedConflicts.value = _unresolvedConflicts.value + response.conflicts
            }

            val pullResponse = api.syncPull(
                since = lastSyncAt?.toInstant()?.toString(),
                collections = null,
            )
            repo.applyServerChanges(pullResponse.items)

            val duration = System.currentTimeMillis() - start
            metrics.record(SyncRecord(
                timestamp = start,
                pushedCount = pushedCount,
                pulledCount = pullResponse.items.size,
                conflictCount = conflictCount,
                errorCount = errorCount,
                durationMs = duration,
                succeeded = true,
            ))

            lastSyncAt = Date()
            _state.value = SyncState.Idle
        }.onFailure { e ->
            val duration = System.currentTimeMillis() - start
            metrics.record(SyncRecord(
                timestamp = start,
                pushedCount = 0,
                pulledCount = 0,
                conflictCount = 0,
                errorCount = 1,
                durationMs = duration,
                succeeded = false,
            ))
            _state.value = SyncState.Error(e.message ?: "Unknown error")
        }
    }

    fun startPolling(intervalMs: Long = 30_000L) {
        stopPolling()
        pollingJob = scope.launch {
            while (isActive) {
                sync()
                delay(intervalMs)
            }
        }
    }

    fun stopPolling() {
        pollingJob?.cancel()
        pollingJob = null
    }

    suspend fun resolve(conflict: SyncConflict, strategy: ResolutionStrategy) {
        runCatching {
            repository?.resolveConflict(conflict, strategy)
            _unresolvedConflicts.value = _unresolvedConflicts.value.filterNot { it.id == conflict.id }
        }.onFailure { e ->
            _state.value = SyncState.Error(e.message ?: "Resolve failed")
        }
    }
}
